crates/libsyntax2/src/lib.rs

```diff
@@ -82,22 +82,68 @@ impl File {
         self.incremental_reparse(edit).unwrap_or_else(|| self.full_reparse(edit))
     }
     pub fn incremental_reparse(&self, edit: &AtomEdit) -> Option<File> {
+        let (node, green, new_errors) =
+            self.reparse_leaf(&edit).or_else(|| self.reparse_block(&edit))?;
+
+        let green_root = node.replace_with(green);
+        let errors = merge_errors(self.errors(), new_errors, node, edit);
+        Some(File::new(green_root, errors))
+    }
+    fn reparse_leaf(&self, edit: &AtomEdit) -> Option<(SyntaxNodeRef, GreenNode, Vec<SyntaxError>)> {
+        let node = algo::find_covering_node(self.syntax(), edit.delete);
+        match node.kind() {
+            | WHITESPACE
+            | COMMENT
+            | DOC_COMMENT
+            | IDENT
+            | STRING
+            | RAW_STRING => {
+                let text = get_text_after_edit(node, &edit);
+                let tokens = tokenize(&text);
+                if tokens.len() != 1 || tokens[0].kind != node.kind() {
+                    return None;
+                }
+
+                let reparser: fn(&mut Parser) = if node.kind().is_trivia() {
+                    // since trivia is omitted by parser when it doesn't have a parent, \
+                    // we need to create one for it
+                    |p| {
+                        p.start().complete(p, ROOT);
+                    }
+                } else {
+                    |p| {
+                        p.bump();
+                    }
+                };
+
+                let (green, new_errors) =
+                    parser_impl::parse_with::<yellow::GreenBuilder>(
+                        &text, &tokens, reparser,
+                    );
+
+                let green = if node.kind().is_trivia() {
+                    green.children().first().cloned().unwrap()
+                } else {
+                    green
```

Comment:
I think it might be easier to construct a trivia green node by hand here? I think ther’s new_leaf constructor for that.

Reply:
That is, to construct by hand without going through green builder and parse with